File: 3504.longest-palindrome-after-substring-concatenation-ii.cpp
#
# @lc app=leetcode id=3504 lang=cpp
#
# [3504] Longest Palindrome After Substring Concatenation II
#
# @lc code=start
class Solution {
public:
    int longestPalindrome(string s, string t) {
        // Calculate character frequencies for both strings
        vector<int> freq_s(26, 0), freq_t(26, 0);
        for (char c : s) freq_s[c - 'a']++;
        for (char c : t) freq_t[c - 'a']++;
        
        // Compute maximum length of palindrome
        int maxLength = 0;
        bool oddCenter = false; // To allow a single odd character in the center if needed
        // Calculate pairs from both sides and check central character possibility
        for (int i = 0; i < 26; ++i) {
            int pairs = min(freq_s[i], freq_t[i]);
            maxLength += pairs * 2;
            
            if ((freq_s[i] > pairs || freq_t[i] > pairs) && !oddCenter) {
                maxLength += 1; // Allow one odd character as center of palindrome
                oddCenter = true;
            }
        }
        return maxLength;
    }
};
# @lc code=end

File: 3504.longest-palindrome-after-substring-concatenation-ii.java
#
# @lc app=leetcode id=3504 lang=java
#
# [3504] Longest Palindrome After Substring Concatenation II
#

# @lc code=start
class Solution {
    // Helper to find the longest palindrome in a single string
    private int longestPalindromeSingle(String str) {
        int n = str.length(), maxLen = 0;
        for (int center = 0; center < n; ++center) {
            // Odd length
            int l = center, r = center;
            while (l >= 0 && r < n && str.charAt(l) == str.charAt(r)) {
                maxLen = Math.max(maxLen, r - l + 1);
                l--; r++;
            }
            // Even length
            l = center; r = center + 1;
            while (l >= 0 && r < n && str.charAt(l) == str.charAt(r)) {
                maxLen = Math.max(maxLen, r - l + 1);
                l--; r++;
            }
        }
        return maxLen;
    }
    
    public int longestPalindrome(String s, String t) {
        int maxLen = 0;
        // Case 1: Palindrome fully in s
        maxLen = Math.max(maxLen, longestPalindromeSingle(s));
        // Case 2: Palindrome fully in t
        maxLen = Math.max(maxLen, longestPalindromeSingle(t));

        // Case 3: Palindrome crossing the boundary
        int n = s.length(), m = t.length();
        // For every possible split (i in s, j in t), expand around boundary
        // Center between s[i] and t[j]
        for (int i = 0; i <= n; ++i) {
            for (int j = 0; j <= m; ++j) {
                // Odd length: center exactly at the boundary
                int l = i - 1, r = j;
                int len = 0;
                while (l >= 0 && r < m && s.charAt(l) == t.charAt(r)) {
                    len += 2; l--; r++;
                }
                maxLen = Math.max(maxLen, len);
                // Even length: center between s[i-1] and t[j]
                // (already handled above, since even/odd is symmetric across boundary)
            }
        }
        // Additionally, single chars at the boundary can form palindromes of len 1
        return Math.max(maxLen, 1);
    }
}
# @lc code=end